spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Attention.scala

```diff
@@ -15,24 +15,58 @@
  */
 package com.intel.analytics.bigdl.nn
 
-import com.intel.analytics.bigdl.Module
-import com.intel.analytics.bigdl.nn.abstractnn.TensorModule
+import javafx.scene.control.Tab
+
+import com.intel.analytics.bigdl._
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity, TensorModule}
 import com.intel.analytics.bigdl.tensor.Tensor
 import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
 
+import scala.language.existentials
 import scala.reflect.ClassTag
 
 /**
  * Implementation of multiheaded attention and self-attention layers.
+ *
  * @param hiddenSize hidden size
  * @param numHeads heads number
  * @param attentionDropout
  */
-class Attention[T: ClassTag](
-  val hiddenSize: Int, val numHeads: Int, val attentionDropout: Float)
- (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+class Attention[T: ClassTag](val hiddenSize: Int, val numHeads: Int, val attentionDropout: Float)
+  (implicit ev: TensorNumeric[T]) extends AbstractModule[Activity, Activity, T] {
+
+  // for prediction
+  private val join1 = nn.JoinTable[T](dimension = 2, nInputDims = -1)
+  private val join2 = nn.JoinTable[T](dimension = 2, nInputDims = -1)
+
+  private val queryLayer = TransformerOperation.dense(
```

Comment:
`join1` is not a good name, make it more meaningful ?

Reply:
ok, changed